libcontainer/cgroups/fs/memory.go

```diff
@@ -55,9 +58,50 @@ func (s *MemoryGroup) Apply(d *cgroupData) (err error) {
 }
 
 func (s *MemoryGroup) SetKernelMemory(path string, cgroup *configs.Cgroup) error {
-	// This has to be done separately because it has special constraints (it
-	// can't be done after there are processes attached to the cgroup).
-	if cgroup.Resources.KernelMemory > 0 {
+	// This has to be done separately because it has special
+	// constraints (it can only be initialized before setting up a
+	// hierarchy or adding a task to the cgroups. However, if
+	// sucessfully initialized, it can be updated anytime afterwards)
+	if cgroup.Resources.KernelMemory != 0 {
+		kmemInitialized := false
+		// Is kmem.limit_in_bytes already set?
+		if b, err := ioutil.ReadFile(filepath.Join(path, "memory.kmem.limit_in_bytes")); err != nil {
+			return err
+		} else {
+			kmemValue, err := strconv.ParseUint(strings.TrimSpace(string(b)), 10, 64)
+			if err != nil {
+				return err
+			}
+			switch system.GetLongBit() {
+			case 32:
+				kmemInitialized = uint32(kmemValue) != uint32(math.MaxUint32)
+			case 64:
+				kmemInitialized = kmemValue != uint64(math.MaxUint64)
+			}
+		}
+
+		if !kmemInitialized {
+			// If hierarchy is set, we can't change the limit
+			if b, err := ioutil.ReadFile(filepath.Join(path, "memory.use_hierarchy")); err != nil {
+				return err
+			} else {
+				val := strings.TrimSpace(string(b))
+				if val != "0" {
+					return fmt.Errorf("Cannot initialize kmem.limit_in_bytes if use_hierarchy is already set")
+				}
+			}
+
+			// If there's already tasks in the cgroup, we can't change the limit either
+			if b, err := ioutil.ReadFile(filepath.Join(path, "tasks")); err != nil {
+				return err
+			} else {
```

Comment:
It's sort of "official" code style :) https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow